Automattic/woocommerce-payments
# Evaluate Psalm for static type checking

[Psalm](https://psalm.dev/) is a static type checker for PHP. It is one of a couple that were mentioned at several talks at the PHP UK conference. I wanted to look into it for a few reasons:

* I'm a fan of strong typing, but there are some blockers to going all out with it.
   * There's a potential runtime penalty of type checking everything (in some cases, and not sure how much)
   * We target a relatively low version of PHP so don't have access to all the features we might want
* We have been looking into how we can ensure Exceptions are properly caught. PHPCS doesn't quite cut it.

So by running static type checking before shipping our code we can work around some of these limitations.

This PR implements Psalm and has a go at fixing a few issues that it raises. At what I think is the strictest ruleset we only raise about 300 errors, which is pretty good since we haven't been developing with this in mind. Psalm is able to figure out the types for around 80% of our code base, which again, is pretty good.

There are more details in the commit messages.

**Testing Instructions**

1. `composer install` - we pull in all the dependencies with composer so there is no need to setup a WordPress testing environment (maybe we can do the same for some of our unit tests later?). The bootstrap file that loads everything is horrible, I hacked it together for the purposes of this experiment.
2. `./vendor/bin/psalm.phar` - gives a relatively verbose output.
3. `./vendor/bin/psalm.phar --output-format=compact` - is a nice PHPCS style output.
4. `./vendor/bin/psalm.phar --report=psalm.txt` - dumps results to file. There are plenty of options here for integrating with CI servers.
5. `./vendor/bin/psalm.phar --stats` - gives details on the typing coverage for each file. It's interesting that where we've made a dedicated class for things, that tends to have 100% coverage.
6. Revert some of the commits making fixes (or make your own new fixes). Re-run the tool and see the results update. Often fixing one thing allows Psalm to understand some typing better, which leads to further errors.

**Next Steps**
* Is this something that people would be interested in getting setup in our repositories?
* Can we do the bootstrap better?
* Can we apply the same bootstrapping to our unit tests?
* There might be additional rules we can apply if we're interested in particular things
* There are some actual bugs this process has turned up, I'll open separate issues for those and get the fixes merged in.
